homeassistant/components/pulseaudio_loopback/switch.py

```diff
@@ -84,47 +86,74 @@ def __init__(self, host, port, buff_sz, tcp_timeout):
         self._buffer_size = int(buff_sz)
         self._tcp_timeout = int(tcp_timeout)
 
-    def _send_command(self, cmd, response_expected):
+    def _send_command(self, cmd, verifier=None):
         """Send a command to the pa server using a socket."""
         sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         sock.settimeout(self._tcp_timeout)
         try:
             sock.connect((self._pa_host, self._pa_port))
             _LOGGER.info("Calling pulseaudio: %s", cmd)
             sock.send((cmd + "\n").encode("utf-8"))
-            if response_expected:
-                return_data = self._get_full_response(sock)
+            if verifier:
+                return_data = self._verify_response(sock, verifier)
                 _LOGGER.debug("Data received from pulseaudio: %s", return_data)
             else:
                 return_data = ""
         finally:
             sock.close()
         return return_data
 
-    def _get_full_response(self, sock):
+    def _verify_response(self, sock, verifier):
         """Get the full response back from pulseaudio."""
         result = ""
-        rcv_buffer = sock.recv(self._buffer_size)
-        result += rcv_buffer.decode("utf-8")
 
-        while len(rcv_buffer) == self._buffer_size:
-            rcv_buffer = sock.recv(self._buffer_size)
-            result += rcv_buffer.decode("utf-8")
+        sock.setblocking(False)
+
+        start_time = monotonic()
+        remaining_timeout = float(self._tcp_timeout)
+
+        while remaining_timeout > 0:
+            ready = select.select([sock], [], [], remaining_timeout)
+            if ready[0]:
+                rcv_buffer = sock.recv(self._buffer_size)
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices